express-api/src/services/keycloak/keycloakService.ts

```diff
@@ -0,0 +1,202 @@
+ * @param   {string}   roleName String name of role in Keycloak
+ * @param   {string}   newRoleName The name to change the role name to.
+ * @returns {IKeycloakRole} The updated role information. Existing role info if cannot be updated.
+ * @throws  {Error} If the newRoleName already exists.
+ */
+const updateKeycloakRole = async (roleName: string, newRoleName: string) => {
+  const roleWithNameAlready: IKeycloakRole = await getRole(newRoleName);
+  // If it already exists, log the error and return existing role
+  if (keycloakRoleSchema.safeParse(roleWithNameAlready).success) {
+    const message = `keycloakService.updateKeycloakRole: Role ${newRoleName} already exists`;
+    logger.warn(message);
+    throw new Error(message);
+  }
+  const response: IKeycloakRole = await getRole(roleName);
+  // Did the role to be changed exist? If not, it will be of type IKeycloakErrorResponse.
+  let role: IKeycloakRole;
+  if (keycloakRoleSchema.safeParse(response).success) {
+    // Already existed. Update the role.
+    role = await updateRole(roleName, newRoleName);
+  } else {
+    // Didn't exist already. Add the role.
+    role = await createRole(newRoleName);
+  }
+
+  // Return role info
+  return role;
+};
+
+// TODO: Complete when user and role services are complete.
+const syncKeycloakUser = async () => {
+  // Does user exist in Keycloak?
+  // Get their existing roles.
+  // Does user exist in PIMS
+  // If user exists in PIMS...
+  // Update the roles in PIMS to match their Keycloak roles
+  // If they don't exist in PIMS...
+  // Add user and assign their roles
+};
+
+interface IKeycloakUsersFilter {
+  lastName?: string;
+  firstName: string; // Currently required. Asked for changes in CSS repo.
+  email?: string;
+  guid?: string;
+}
+/**
+ * @description Retrieves Keycloak users based on the provided filter.
+ * @param {IKeycloakUsersFilter} filter The filter to apply when retrieving users.
+ * @returns {IKeycloakUser[]} A list of Keycloak users.
+ */
+const getKeycloakUsers = async (filter: IKeycloakUsersFilter) => {
+  // Get all users from Keycloak for IDIR
+  // CSS API returns an empty list if no match.
+  let users: IKeycloakUser[] = ((await getIDIRUsers(filter)) as IKeycloakUsersResponse).data;
+  // Add BCeID if GUID was included.
+  if (filter.guid) {
+    users = users.concat(((await getBothBCeIDUser(filter.guid)) as IKeycloakUsersResponse).data);
+  }
+  // Return list of users
+  return users;
+};
+
+/**
+ * @description Retrieves a Keycloak user that matches the provided guid.
+ * @param {string} guid The guid of the desired user.
+ * @returns {IKeycloakUser} A single Keycloak user.
+ * @throws If the user is not found.
+ */
+const getKeycloakUser = async (guid: string) => {
+  // Should be by guid. Only way to guarantee uniqueness.
+  const user: IKeycloakUser = (await getKeycloakUsers({ guid: guid, firstName: '' })).at(0); // TODO: Remove firstName after fix is applied to package.
```

Comment:
This would probably result in an exception if getKeycloakUsers failed for some reason, doing `?.at(0);` should be enough.